components/rendering/src/markdown.rs

```diff
@@ -1,20 +1,121 @@
 
 use pulldown_cmark as cmark;
 use self::cmark::{Parser, Event, Tag, Options, OPTION_ENABLE_TABLES, OPTION_ENABLE_FOOTNOTES};
 use slug::slugify;
 use syntect::easy::HighlightLines;
 use syntect::html::{start_coloured_html_snippet, styles_to_coloured_html, IncludeBackground};
 
-use errors::Result;
+use errors::{Result, Error};
 use utils::site::resolve_internal_link;
 use context::Context;
 use highlighting::{SYNTAX_SET, THEME_SET};
 use short_code::{SHORTCODE_RE, ShortCode, parse_shortcode, render_simple_shortcode};
 use table_of_contents::{TempHeader, Header, make_table_of_contents};
 
+struct GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    context: &'a Context<'a>,
+    errors: &'c mut Vec<Error>,
+    parser: Peekable<Parser<'b>>,
+}
+
+fn is_internal_link(link: &str) -> bool {
+    link.starts_with("./")
+}
+
+impl<'a, 'b, 'c> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    fn new(context: &'a Context<'a>, parser: Parser<'b>, errors: &'c mut Vec<Error>) -> GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+        GutenbergFlavoredMarkdownParser {
+            context: context,
+            errors: errors,
+            parser: parser.peekable()
+        }
+    }
+
+    fn push_error<E: Into<Error>>(&mut self, error: E) {
+        self.errors.push(error.into());
+    }
+
+    fn empty_event(&self) -> Event<'b> {
+        Event::Html(Owned("".to_string()))
+    }
+
+    fn process_link(&mut self, link: Tag<'b>) -> Event<'b> {
+        match link {
+            Tag::Link(ref href, ref text) => {
+                if is_internal_link(href) {
+                    match resolve_internal_link(href, self.context.permalinks) {
+                        Ok(url) => {
+                            Event::Start(Tag::Link(Owned(url), text.clone()))
+                        },
+                        Err(_) => {
+                            self.push_error(format!("Relative link {} not found.", href));
+                            self.empty_event()
+                        }
+                    }
+                } else{
+                    Event::Start(Tag::Link(href.clone(), text.clone()))
+                }
+            },
+            _ => unreachable!()
+        }
+    }
+
+    fn add_id_to_header(&mut self, num: i32) -> Event<'b> {
+        Event::Start(Tag::Header(num))
+    }
+
+    // TODO: Find a better name for this function.
+    fn process_event(&mut self, event: Event<'b>) -> Event<'b> {
+        match event {
+            Event::Start(link @ Tag::Link(_, _)) => self.process_link(link),
+            Event::Start(Tag::Header(num)) => self.add_id_to_header(num),
+            _ => event,
+        }
+    }
+}
 
+impl<'a, 'b, 'c> Iterator for GutenbergFlavoredMarkdownParser<'a, 'b, 'c> {
+    type Item = Event<'b>;
+
+    fn next(&mut self) -> Option<Event<'b>> {
+        match self.parser.next() {
+            Some(event) => Some(self.process_event(event)),
+            None => None,
+        }
+    }
+}
+
+#[allow(unused_variables)]
```

Comment:
is that just while it's a wip?

Reply:
yes, to make fewer warnings.  We'll make sure to remove these and fix warnings before finally submitting.